rskj-core/src/main/java/co/rsk/db/JournalTrieCache.java

```diff
@@ -0,0 +1,763 @@
+        firstLevelCache.getDeletedAccounts().forEach(account -> trie.deleteRecursive(account.getData()));
+        firstLevelCache.getUpdatedKeys().forEach(key -> {
+            byte[] value = firstLevelCache.get(key);
+            trie.put(key, value);
+        });
+        // clear the complete cache
+        firstLevelCache.clear();
+    }
+
+    /**
+     * Insure that there is no changes in cache anymore (ie all changes have been committed or rolled back)
+     */
+    private void assertNoCache() {
+        if (!cache.isEmpty()) {
+            throw new IllegalStateException();
+        }
+    }
+
+    /**
+     * A class to implement a multilevel cache system, so that the same
+     *  object can record changes at different caching level
+     */
+    private static class MultiLevelCache {
+
+        public static final int FIRST_CACHE_LEVEL = 1;
+        private int depth = FIRST_CACHE_LEVEL;
+        private final Map<ByteArrayWrapper, Map<ByteArrayWrapper, Map<Integer, byte[]>>> valuesPerLevelPerKeyPerAccount
+                = new HashMap<>();
+        private final Map<Integer, Set<ByteArrayWrapper>> deletedAccountsPerLevel = new HashMap<>();
+        private final Map<Integer, Set<ByteArrayWrapper>> updatedKeysPerLevel = new HashMap<>();
+
+        MultiLevelCache() {
+            initLevel(depth);
+        }
+
+        int getDepth() {
+            return depth;
+        }
+
+        int getNextLevel() {
+            depth++;
+            initLevel(depth);
+            return depth;
+        }
+
+        void put(ByteArrayWrapper key, byte[] value, int cacheLevel) {
+            ByteArrayWrapper accountWrapper = getAccountWrapper(key);
+            // if this account is in the deletedAccounts list, dont remove it, because this means that the lower
+            // level is deleted
+            Map<ByteArrayWrapper, Map<Integer, byte[]>> valuesPerLevelPerKey
+                    = this.valuesPerLevelPerKeyPerAccount.computeIfAbsent(accountWrapper, k -> new HashMap<>());
+            Map<Integer, byte[]> valuesPerLevel = valuesPerLevelPerKey.computeIfAbsent(key, k -> new HashMap<>());
+            valuesPerLevel.put(cacheLevel, value);
+            updatedKeysPerLevel.get(cacheLevel).add(key);
+        }
+
+        byte[] get(ByteArrayWrapper key, int cacheLevel) {
+            ByteArrayWrapper accountWrapper = getAccountWrapper(key);
+            Map<ByteArrayWrapper, Map<Integer, byte[]>> valuesPerLevelPerKey = valuesPerLevelPerKeyPerAccount.get(accountWrapper);
+            if (valuesPerLevelPerKey != null) {
+                Map<Integer, byte[]> valuesPerLevel = valuesPerLevelPerKey.get(key);
+                if (valuesPerLevel != null) {
+                    return valuesPerLevel.get(cacheLevel);
+                }
+            }
+            // TODO Prefer Optional<T> over null
+            return null;
+        }
+
+        byte[] getNewestValue(ByteArrayWrapper key, int maxLevel) {
```

Comment:
This is not O(1). This is O(L) where L is the number of levels the value has been modified. This is no different with the current multi-level cache approach.
The spec required a cache based on single last state which could be checked in O(1). All changes should go to a log.

Reply:
OK. I've been focused on having the same structure (instance) to store data across the different cache levels, and I miss the point regarding the complexity class of the get method.
I look into it to upgrade my PR

Reply:
@SergioDemianLerner 
I've just upgraded my PR to take into account your comment.
Now the get accesses to the cache are O(1) as requested